### PR TITLE
[Bluetooth][Non-ACR] Fix Native callback GC issue

### LINKED ISSUE #1671: [Bluetooth] Native Callback GC issue
https://github.com/Samsung/TizenFX/blob/98dcbeeefd4172dbe9fab93fafe101200ce6255e/src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs#L119

https://github.com/Samsung/TizenFX/blob/98dcbeeefd4172dbe9fab93fafe101200ce6255e/src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs#L217

https://github.com/Samsung/TizenFX/blob/98dcbeeefd4172dbe9fab93fafe101200ce6255e/src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs#L238

Native Callback could be released by GC before completed async operation

### PR DESCRIPTION
**Pull request by Wootak**
fix #1671 

Signed-off-by: Wootak Jung <wootak.jung@samsung.com>

### CODE REVIEW

File: src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs [hunk 1]
```diff
@@ -115,20 +118,37 @@ internal void SendNotification(BluetoothGattCharacteristic characteristic, strin
...
+            TaskCompletionSource<bool> task = new TaskCompletionSource<bool>();
+            int requestId = 0;
+
+            lock (this)
+            {
+                requestId = _requestId++;
+                _sendIndicationTaskSource[requestId] = task;
+            }
+
+            _sendIndicationCallback = (result, address, serverHandle, characteristicHandle, completed, userData) =>
```
Comment (myroot) [inline src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs:130]: `_sendIndicationCallback` is overwritten by following request

Comment (myroot) [reply src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs:130]: Please refer this PR https://github.com/Samsung/TizenFX/pull/1674
it is same case

Comment (Wootak) [reply src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs]: Done.

File: src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs [hunk 2]
```diff
@@ -113,22 +119,42 @@ internal void SendNotification(BluetoothGattCharacteristic characteristic, strin
...
             {
-                _notificationSent?.Invoke(characteristic, new NotificationSentEventArg(server, address, result, completed));
+                _notificationSent?.Invoke(characteristicHandle, new NotificationSentEventArg(_sendIndicationServer[requestId], clientAddress, result, completed));
                 if (completed)
                 {
-                    tcs.SetResult(true);
+                    _sendIndicationTaskSource[requestId].SetResult(true);
                 }
-            };
+                _sendIndicationTaskSource.Remove(requestId);
```
Comment (myroot) [inline src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs:132]: When remove a `BluetoothGattServer` from _sendIndicationServer?

File: src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs [hunk 3]
```diff
@@ -113,22 +119,42 @@ internal void SendNotification(BluetoothGattCharacteristic characteristic, strin
...
-            Interop.Bluetooth.BtGattServerNotificationSentCallback cb = (result, address, serverHandle, characteristicHandle, completed, userData) =>
+            int requestId = (int)userData;
+            if (_sendIndicationTaskSource.ContainsKey(requestId))
             {
-                _notificationSent?.Invoke(characteristic, new NotificationSentEventArg(server, address, result, completed));
+                _notificationSent?.Invoke(characteristicHandle, new NotificationSentEventArg(_sendIndicationServer[requestId], clientAddress, result, completed));
                 if (completed)
                 {
-                    tcs.SetResult(true);
+                    _sendIndicationTaskSource[requestId].SetResult(true);
```
Comment (myroot) [inline src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs:130]: not completed, what is means?
If you not set a result on TaskCompletetionSource, a task is never completed

File: src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs [hunk 4]
```diff
@@ -113,22 +119,42 @@ internal void SendNotification(BluetoothGattCharacteristic characteristic, strin
...
-        internal Task<bool> SendIndicationAsync(BluetoothGattServer server, BluetoothGattCharacteristic characteristic, string clientAddress)
+        void SendIndicationCallback(int result, string clientAddress, IntPtr serverHandle, IntPtr characteristicHandle, bool completed, IntPtr userData)
         {
-            TaskCompletionSource<bool> tcs = new TaskCompletionSource<bool>();
-            Interop.Bluetooth.BtGattServerNotificationSentCallback cb = (result, address, serverHandle, characteristicHandle, completed, userData) =>
+            int requestId = (int)userData;
+            if (_sendIndicationTaskSource.ContainsKey(requestId))
             {
-                _notificationSent?.Invoke(characteristic, new NotificationSentEventArg(server, address, result, completed));
+                _notificationSent?.Invoke(characteristicHandle, new NotificationSentEventArg(_sendIndicationServer[requestId], clientAddress, result, completed));
```
Comment (myroot) [inline src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs:127]: I recommend to change BluetoothGattServer like below 
``` c#
        private BluetoothGattServer()
        {
            _impl = new BluetoothGattServerImpl();
            _impl._notificationSent += (s, e) => NotificationSend?.Invoke(this, e);
        }

        /// <summary>
        /// (event) This event is called when the indication acknowledgement is received for each notified client.
        /// </summary>
        /// <since_tizen> 3 </since_tizen>
        public event EventHandler<NotificationSentEventArg> NotificationSent;
```

File: src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs [hunk 5]
```diff
@@ -27,9 +27,15 @@ internal class BluetoothGattServerImpl
     {
         private BluetoothGattServerHandle _handle;
         internal event EventHandler<NotificationSentEventArg> _notificationSent;
+        int _requestId = 0;
+        Dictionary<int, TaskCompletionSource<bool>> _sendIndicationTaskSource = new Dictionary<int, TaskCompletionSource<bool>>();
+        Dictionary<int, BluetoothGattServer> _sendIndicationServer = new Dictionary<int, BluetoothGattServer>();
```
Comment (myroot) [inline src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs:32]: Remove unused variable

File: src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs [hunk 6]
```diff
@@ -113,22 +119,47 @@ internal void SendNotification(BluetoothGattCharacteristic characteristic, strin
...
-        internal Task<bool> SendIndicationAsync(BluetoothGattServer server, BluetoothGattCharacteristic characteristic, string clientAddress)
+        void SendIndicationCallback(int result, string clientAddress, IntPtr serverHandle, IntPtr characteristicHandle, bool completed, IntPtr userData)
         {
-            TaskCompletionSource<bool> tcs = new TaskCompletionSource<bool>();
-            Interop.Bluetooth.BtGattServerNotificationSentCallback cb = (result, address, serverHandle, characteristicHandle, completed, userData) =>
+            int requestId = (int)userData;
+            if (_sendIndicationTaskSource.ContainsKey(requestId))
             {
-                _notificationSent?.Invoke(characteristic, new NotificationSentEventArg(server, address, result, completed));
+                _notificationSent?.Invoke(characteristicHandle, new NotificationSentEventArg(_sendIndicationServer[requestId], clientAddress, result, completed));
```
Comment (myroot) [inline src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs:127]:
```suggestion
                _notificationSent?.Invoke(characteristicHandle, new NotificationSentEventArg(this, clientAddress, result, completed));
```

Comment (Wootak) [reply src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs]: Thank you!

Comment (Wootak) [reply src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs:127]: error occured. 'this' indicates GattServerImpl not GattServer.

Tizen.Network.Bluetooth/BluetoothGattImpl.cs(126): CS1503: Argument 1: cannot convert from 'Tizen.Network.Bluetooth.BluetoothGattServerImpl' to 'Tizen.Network.Bluetooth.BluetoothGattServer'

Comment (myroot) [reply src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs:127]: Why sender is `characteristicHandle`?

Comment (myroot) [reply src/Tizen.Network.Bluetooth/Tizen.Network.Bluetooth/BluetoothGattImpl.cs:127]: How about change like below?
``` c#
    public class NotificationSentEventArg : EventArgs
    {
        internal NotificationSentEventArg(BluetoothGattServer server, string clientAddress, int result, bool completed)
        {
            Result = result;
            ClientAddress = clientAddress;
            Server = server;
            Completed = completed;
        }
        public BluetoothGattServer Server { get; internal set;}
        public string ClientAddress { get; }
        public int Result { get; }
        public bool Completed { get; }
    }
...
       _notificationSent?.Invoke(characteristicHandle, new NotificationSentEventArg(null, clientAddress, result, completed));
...

        private BluetoothGattServer()
        {
            _impl = new BluetoothGattServerImpl();
            _impl._notificationSent += (s, e) => { 
                    e.Server = this;
                    NotificationSend?.Invoke(this, e); 
            };
        }

```

**Comment by myroot** [comment]
@Wootak  Please also submit on API7

**Comment by Wootak** [comment]
@myroot I makde a PR on API7 also. Thank you.
https://github.com/Samsung/TizenFX/pull/1699

**Pull request merged**